Fix bugs, add tests, add predefined locales, change behavior of formatInt

lovasoa opened the pull request:
I needed these changes for a project, so I already published them:

http://package.elm-lang.org/packages/lovasoa/elm-format-number/latest/FormatNumber

I'm open for discussion on these changes.

lovasoa committed:
add package.json

lovasoa committed:
Fix bug with decimals=0

lovasoa committed:
reset version

lovasoa committed:
do not print decimals with it

lovasoa committed:
Refactoring

Refactor, add tests, fix bugs

lovasoa committed:
Fix bug and add tests

lovasoa committed:
Add locales

lovasoa committed:
version bump

lovasoa committed:
version bump

lovasoa committed:
Update README

lovasoa committed:
Fix doc

lovasoa committed:
Add travic CI

cuducos commented on README.md:
```diff
@@ -1,7 +1,12 @@
 # Elm Format Number
 
+[![Build Status](https://travis-ci.org/lovasoa/elm-format-number.svg?branch=master)](https://travis-ci.org/lovasoa/elm-format-number)
```
If you rename `lovasoa -> cuducos` I can merge directly to `master`, otherwise I'll merge to a secondary branch, fix it and than bring to `master` (no problem, BTW). The same for `package.json` and `elm-package.json`.

cuducos commented on README.md:
```diff
@@ -15,20 +20,6 @@ defaultLocale =
     , thousandSeparator = ","
     , decimalSeparator = "."
     }
-
```
I wouldn't remove the examples of how to define you own `Locales`. Would you mind bringing them back?

cuducos commented on README.md:
````diff
@@ -37,12 +28,10 @@ yetAnotherLocale =
 formatFloat : Locale -> Float -> String
 ```
 
-Format a `Float` number as a pretty string. For example, using the locales defined above:
````
If there are two `Locale`s above get back, keep the plural ; )

cuducos commented on README.md:
````diff
@@ -37,12 +28,10 @@ yetAnotherLocale =
 
 ```elm
 formatFloat defaultLocale 1234.5567 == "1,234.56"
-formatFloat alternativeLocale -7654.3210 == "-7.654,321"
````
Why remove edge case examples? [Users might ask themselves](https://www.reddit.com/r/elm/comments/5jcz59/just_wrote_an_elm_package_to_format_numbers/dbffbw8/) _does this work with negative numbers?_, _does it round or just strip digits off?_…

lovasoa replied:
I wanted to make the README a very short description, that gives an quick overview of the project, and not a documentation.

The full documentation is inside the module. I actually added a lot of edge cases to the documentation.

cuducos replied:
Ok, that's really good you added a lot of edge cases, I mean it ❤️ 

But having it _half-documented_ in the `README.md` and fully documented elsewhere leads to confusion. I'm OK with:

1. having it all on the `README.md` and fully duplicated at Elm Packages
1. **or** having everything at Elm Packages (and `README.md` just points there, as you suggested in this PR).

But anywhere in-between leads to confusion IMHO. Let's go for option number 2?

rogeriochaves replied:
+1 for having links

cuducos commented on README.md:
````diff
@@ -51,11 +40,10 @@ formatFloat yetAnotherLocale -0.01 == "0.0"
 formatInt : Locale -> Int -> String
 ```
 
-Format a `Int` number as a pretty string. For example, using the locales defined above:
````
If are two `Locale`s above get back, keep the plural ; )

cuducos commented on README.md:
````diff
@@ -51,11 +40,10 @@ formatFloat yetAnotherLocale -0.01 == "0.0"
 
 ```elm
-formatInt defaultLocale 0 == "0.00"
-formatInt defaultLocale 1234567890 == "1,234,567,890.0"
````
Why remove a basic example?

cuducos commented on elm-package.json:
```diff
@@ -1,7 +1,7 @@
 {
-    "version": "2.0.0",
```
Is this a major update? Does it need to change compatibility?

lovasoa replied:
It's an error on my side :face_with_head_bandage: sorry !

cuducos replied:
No problem… Minor minor minor details in a lot of coding ; )

rogeriochaves replied:
if you ran `elm package bump` elm probably told you to go with 1.x.x since you changed the repo field and doesn't have permission to publish as cuducos. Maybe that was what happened (:

cuducos commented on src/FormatNumber.elm:
```diff
@@ -1,20 +1,42 @@
 
 @docs formatFloat, formatInt
 
+# Known bugs
```
Awesome! Many thanks for that 🎉

cuducos commented on src/FormatNumber.elm:
```diff
@@ -1,20 +1,42 @@
+ * https://github.com/elm-lang/elm-compiler/issues/264
+ * https://github.com/elm-lang/elm-compiler/issues/1246
+
+This library won't work with large numbers (over 2^31) until elm itself is fixed
```
Total 💅 but… would you mind using _Elm_ instead of _elm_ and adding a `:` at the end of the sentence. Feel free to call me names if you want ; )

cuducos commented on src/FormatNumber.elm:
```diff
@@ -26,6 +48,42 @@ type alias Locale =
 
 
+
+-- Locales from
```
Personally I don't like to fix `Locale` because it does not depends _only_ of the region/culture/locale: For example in Brazil we use what you and Oracle called `spanishLocale`, **but**:

* we use _one_ decimal when it comes to measuring certain lengths (e.g: my home is 1,2km far from the park)
* we use _two_ decimals when it comes to measuring other lengths (e.g.: I'm 1,68m tall)
* we use _two_ decimals for almost all prices (e.g: a beer costs around R$ 2,10 at the market)
* but certain prices comes with _three_ decimals (e.g. gas costs R$ 3,599/L, or stocks for a given company costs R$ 10,324)

So even if `spanishLocale` is what makes sense in Brazil, I would only be able to use it in one of the four situations described there.

I'd strongly advice to let users decide about their own locales. Maybe your `Locale`s could be part of the documentation, as examples, but not as part of the package source code IMHO.

lovasoa replied:
I didn't remove the `Locale` constructor. I simply added a few predefined ones for ease of use.

The predefined spanishLocale would actually be usefull even in the cases you describe. You would do:

```elm
lengths = { spanishLocale | decimals = 1 }
heights = { spanishLocale | decimals = 2 }
```
And it would be very readable :+1: 

Plus, if a new major versions added fields in `Locale`, old code would still work.

cuducos replied:
Totally agree. You got me 💃 🕺 

Can you add these bits to the documentation? Your idea is good, let's make it more evident! (Or is it already there and I missed it?)

cuducos commented on src/FormatNumber.elm:
```diff
@@ -37,115 +95,61 @@ type alias Locale =
 
     >>> formatInt { decimals = 1, thousandSeparator = ",", decimalSeparator = "." } 0
-    "0.0"
+    "0"
```
If we **explicitly** ask for _1 decimal_ (`decimals = 1` within the `Locale`), this _should_ be `0.0`, not `0`. However maybe it make sense to add an option for the `Locale`, something like `ignoreZeroDecimals : Bool` that controls whether  the format function can skip decimals when they're equal to zero. What do you think?

lovasoa replied:
The user doesnt explicitly asked for a number of decimals. It's just part of the locale.

My opinion is that integers should never be printed with decimals. An int doesn't **have** decimals. It's a different mathematical object. If an user ends up wanting to print an `Int` with decimals, then he should have been using a `Float` from the beginning.

This allows users to use a single locale for int and floats, instead of having to define all their locals twice (for int with `decimals=0` and for float with a different number of decimals).

cuducos replied:
> The user doesnt explicitly asked for a number of decimals. It's just part of the locale.

It's part of the `Locale` user coded or choose, ie. a choice.

> My opinion is that integers should never be printed with decimals

I totally disagree. Prices are usually expressed with cents (or pences, or whatever), for example. Formally speaking it's _$ 1,00_, not _$ 1_ (even if informally we use _$ 1_). I still believe adding an option in the `Locale` is a better choice.

rogeriochaves replied:
Hmmmmm

To avoid confusion, should we remove `formatInt` completely? We can have a `format` function that just take floats to simplify the lib and avoid confusions. If the user wants she/he can convert their int to float.

Same way as `text` in html only accepts strings, and I always have to convert my stuff to string first.

lovasoa replied:
If your program  manipulates amounts of money, it has to be done with either a float or an int representing **the cents**. I can't think of any good reason one would store an amount of money in dollars as an `Int`.

So an user would use the API this way:

```elm
type alias Cents = Int

displayDollars : Cents -> String
displayDollars cents = "$ " ++ (formatFloat {usLocale | decimals = 2} (toFloat(cents)/100))
```

cuducos replied:
It was just an example. Conceptually speaking I still believe if user specify decimals (eg `decimals = 1` within the `Locale`), this should be `0.0`, not `0` **unless** the user specify that the formatter can skip zeros. That's why I'm in favor of adding an option to the `Locale`.

cuducos replied:
> To avoid confusion, should we remove formatInt completely?

Great idea, @rogeriochaves… maybe not in this PR, but later surely I'll take that ; )

lovasoa replied:
@rogeriochaves But formatting integers is sometimes needed. I think it's great to have only one locale, and being able to use it to format both floats and ints (that don't have decimals).

cuducos replied:
Putting `Int` and `Float` into the screen is needed. We use `my_number |> toString |> Html.text` regularly for that ; )

lovasoa replied:
Yes, and being able to to 

```elm
my_number |> formatInt frenchLocale |> Html.text
```

would be great.

cuducos replied:
Nothing that causes confusion is great.

cuducos commented on src/FormatNumber.elm:
```diff
@@ -37,115 +95,61 @@ type alias Locale =
+        LT ->
+            formatInt locale (-num) |> String.cons '-'
+
+        EQ ->
```
Same as above: if users explicitly asked for decimals, we can't just ignore that.

lovasoa replied:
The user explicitly gave an Int :wink:

cuducos replied:
> The user explicitly gave an Int 

Nope. The user explicitly passed an `Int` to a function that format it as `String` using a `Locale` that has a `decimals` field.

If user just wanted to print an `Int` it would be way easier to just pass `Int` to something else: e.g. `toString myIntegerNumber`.

cuducos commented on src/FormatNumber.elm:
```diff
@@ -37,115 +95,61 @@ type alias Locale =
-        |> formatFloat locale
+    case compare num 0 of
+        LT ->
+            formatInt locale (-num) |> String.cons '-'
```
Why do we need a special treatment for negatives now? It's not clear… was it broken before?

lovasoa replied:
Because we don't use `toString` to format numbers anymore, so we have to add the minus sign ourselves.

I removed all uses of `toString` on arbitrary numbers, because the old library assumed that toString simply prints the decimals of numbers, whereas it is not always the case.

```elm
> toString 0.0000001
"1e-7" : String
```

Maybe I should add a test case with this precise example...

cuducos replied:
Great improvement. Many many thanks 🎊

rogeriochaves replied:
+1 for adding a test case

cuducos replied:
> +1 for adding a test case

Would you mind adding that, @lovasoa? Otherwise I can add it to my TODO list and commit after merging your PR, no worries…

cuducos commented on src/Helpers.elm:
```diff
@@ -0,0 +1,55 @@
+    >>> nDigits 2 0.10
+    "10"
+-}
+
```
Can we delete the lines between the comments/test and the function?

cuducos commented on src/Helpers.elm:
```diff
@@ -0,0 +1,55 @@
+-}
+
+
+nDigits : Int -> Float -> String
```
IMHO `nDigits` is not a good name: it's not the number of digits (my first thought for the `n` prefix to the word `Digits`). What about just `digits`?

cuducos commented on src/Helpers.elm:
```diff
@@ -0,0 +1,55 @@
+    >>> splitIntRec 12345 []
+    ["12", "345"]
+-}
+splitIntRec : Int -> List String -> List String
```
This function has a terrible API. It always needs a empty list in the first call (why? the meaning of that is not clear , you can't guess by the function signature `splitIntRec : Int -> List String -> List String`; BTW the signature made me believe it was about the `List String`, not about the the `Int`)… Also I took a while to realize `splitIntRec` was `splitIntRecursively`, but that might be just me.

I strongly believe my old `addThousandSeparator` was way longer but it's signature and code was significantly more readable and meaningful.

If you really wanna stick to your solution, I'd explore the `List.foldr` or `List.foldl` functions instead.

lovasoa replied:
Yes, you are right, the function name is bad.
And yes, the API is uneasy. This is a helper recursive function for formatInt. It is not exposed to users. I'm open to propositions to make it easier to understand.
I don't see why you would need a fold here, however.

cuducos replied:
>  It is not exposed to users.

I'm not taking into account final users. Contributors (like future-me and future-you) might also struggle with internal APIs. I'll try to refactor it and add here as a suggestion later today ; )

cuducos replied:
What about this:

```elm
import List.Split exposing (chunksOfRight)
import List


split : Int -> List String
split number =
    number
        |> toString
        |> String.toList
        |> List.map String.fromChar
        |> chunksOfRight 3
```

We'll have to add [`List.Split`](https://github.com/elm-community/list-split) as a dependency, but I think it worth it!

cuducos replied:
@lovasoa's said in the comments:

> As I explained, we can't just use toString. toString gives no guarantee that the result is the sequence of digits representing the number. 

Is that the case for `Int` too? This is a sincere questions (your example used `0.000001` so I thought it would be a problem with `Float`).

cuducos replied:
What about, I postponed the `toString` to the very last steps:

```elm
toReversedList : Int -> List Int
toReversedList number =
    if number < 10 then
      [ number ]
    else
        let
            remainder : Int
            remainder =
                floor (toFloat number / 10)
        in
            number % 10 :: toReversedList remainder


split : Int -> List String
split number =
    number
        |> toReversedList
        |> List.reverse
        |> chunksOfRight 3
        |> List.reverse
        |> List.map (\l -> List.map toString l)
        |> List.map (\l -> String.join "" l)
```

cuducos reviewed (changes requested):
Hi @lovasoa,

Many thanks for your suggestions. I really liked many of them. In such a extended PR (conceptually speaking) however, I tend to disagree with you mainly when it comes to two things:

* I'm against default locales by country/region/culture because number formatting (specially the amount of decimals) depends mostly on the use case not on the country/region/culture
* I'm agains dropping the decimals without making it explict — in your code, even with `decimals = 2`, for examples, some numbers will return without decimals; thus I argued for an option to make this behavior explicit.

I expanded both points on inline comments.

lovasoa commented:
As I explained, we can't just use toString. toString gives no guarantee that the result is the sequence of digits representing the number. 

25 января 2017 г. 20:39:31 GMT+03:00, Eduardo Cuducos <notifications@github.com> пишет:
>cuducos commented on this pull request.
>
>
>
>> +
>+        fint =
>+            (round (f * multiplicator))
>+    in
>+        splitIntRec fint []
>+            |> String.concat
>+            |> String.right digits
>+            |> String.padLeft digits '0'
>+
>+
>+{-| Recursive helper to format an integer
>+
>+    >>> splitIntRec 12345 []
>+    ["12", "345"]
>+-}
>+splitIntRec : Int -> List String -> List String
>
>What about this:
>
>```elm
>import List.Split exposing (chunksOfRight)
>import List
>
>
>split : Int -> List String
>split number =
>    number
>        |> toString
>        |> String.toList
>        |> List.map String.fromChar
>        |> chunksOfRight 3
>```
>
>We'll have to add
>[`List.Split`](https://github.com/elm-community/list-split) as a
>dependency, but I think it worth it!
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/cuducos/elm-format-number/pull/2

-- 
Простите за краткость, создано в K-9 Mail.

lovasoa commented:
Ok, let's do like that. This gives the advantage of never having out of date comments, as comments in the Elm files are tested with elm doc test. 

25 января 2017 г. 15:46:47 GMT+03:00, Eduardo Cuducos <notifications@github.com> пишет:
>cuducos commented on this pull request.
>
>
>
>>  
> ```elm
> formatFloat defaultLocale 1234.5567 == "1,234.56"
>-formatFloat alternativeLocale -7654.3210 == "-7.654,321"
>
>Ok, that's really good you added a lot of edge cases, I mean it ❤️ 
>
>But having it _half-documented_ in the `README.md` and fully documented
>elsewhere leads to confusion. I'm OK with:
>
>1. having it all on the `README.md` and fully duplicated at Elm
>Packages
>1. **or** having everything at Elm Packages (and `README.md` just
>points there, as you suggested in this PR).
>
>But anywhere in-between leads to confusion IMHO. Let's go for option
>number 2?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/cuducos/elm-format-number/pull/2

-- 
Простите за краткость, создано в K-9 Mail.

lovasoa commented:
That would be okay. And you still could replace 10 by 1000 in your implementation, and use String.padLeft, and you wouldn't need to reverse the string twice neither to use an external library. 

The principle you use here is the same as the one I used. We could simply find a better name for the recursive helper function.

lovasoa committed:
Rename nDigits to digits

See https://github.com/cuducos/elm-format-number/pull/2#discussion_r97743595

lovasoa committed:
Rename splitIntRec to splitThousands...

... and improve its API

See: https://github.com/cuducos/elm-format-number/pull/2#discussion_r97744598

lovasoa committed:
Rewrite the README

Remove documentation from the README

lovasoa committed:
Add documentation and tests to the main module

cuducos commented on src/Helpers.elm:
```diff
@@ -36,22 +36,29 @@ digits digits f =
+            else
+                (toString remainingNumber) :: accumulator
+    in
+        splitRemaining number []
```
🍾 Wow! Really good refactor, many thanks.

I played around with it a bit more, yet trying to enhance the API. What are your thoughts?

```elm
splitThousands : Int -> List String
splitThousands number =
    if number >= 1000 then
        let
            remaining : List String
            remaining =
                splitThousands (number // 1000)

        in
            [ number % 1000 ]
                |> List.map toString
                |> List.map (String.padLeft 3 '0')
                |> List.append remaining
                
    else
        [ toString number ]
```

Once we agree about this function, I merge it!

(BTW, thanks for b31d9fe, it's really good!)

lovasoa replied:
Is there a reason for converting to and from Float instead of simply using integer division ?
Plus, you need to pad numbers with 0's.

lovasoa replied:
Plus, my implementation is tail-recursive, while yours isn't.

cuducos replied:
Good points! Just edited the previous comment updating with your suggestions ; )

cuducos replied:
Way better BTW ; )

lovasoa replied:
This one looks great, but it's still not tail-recursive.

cuducos replied:
> This one looks great, but it's still not tail-recursive.

I guess this is not a big issue for formatting numbers, is it? I'd say _readability counts_ and _practicality beats purity_, but as I come from Python there a lots of bias in scene ; )

lovasoa commented:
Plus, I wanted to use specialized unicode characters instead of the traditional space and comma, and to add a locale for arabic. What do you think about that ?

lovasoa committed:
Use a thin non-breakable space in frenchLocale

cuducos commented:
> Plus, I wanted to use specialized unicode characters instead of the traditional space and comma, and to add a locale for arabic. What do you think about that ?

That would be awesome 👍 

@fcevado said one of these days that he would like to code that too, but I guess he hasn't started yet. Anyway you both could talk, discuss ideas and implementations if you like ; )

fcevado commented:
@cuducos I was waiting to merge this PR to start, so I would fork the updated master 😛 

I think @lovasoa is talking about another default locale... We were talking about the thousands separation(that is different for some kind of chinese representation and the indian/pakistani representation). That would have some impact in the `Helper` module.

lovasoa commented:
@fcevado adding a new locale for arabic means adding new fields to the Locale type (something like `numerals`, for eastern arabic numerals). It would be nice to make a list of all the fields that would be needed for a real internationalization of this module.

fcevado commented:
@lovasoa I will search for official documentation for those patterns, to come up with some parameters for the locale.

but I talked to @cuducos  about having some info about the numeric system that is being representated... the locale would be something like that
```elm
type alias Locale =
    { decimals : Int
    , thousandSeparator : String
    , decimalSeparator : String
    , numericSystem : Arabic | Chinese | Indian
    }
```
but i think more info is needed to define it correctly.

lovasoa commented:
We could have an array of `Char` (or even `String`) as numericSystem, and export `arabicNumericSystem`, `chineseNumericSystem`, and `indianNumericSystem`. That would be more expressive and no more complex to use.

cuducos commented:
I really don't want to stop you from discussing, but… IF you open an Issue about probably it would be easier for you both to get back to these messages later (and it'd allow more people interested in non-Latin numeric system to join you)…

lovasoa commented:
Maybe decimals shouldn't be part of Locale, then...

27 января 2017 г. 16:56:46 GMT+03:00, Eduardo Cuducos <notifications@github.com> пишет:
>cuducos commented on this pull request.
>
>
>
>>  
> 
> {-| Format a integer number as a pretty string:
> 
>>>> formatInt { decimals = 1, thousandSeparator = ",", decimalSeparator
>= "." } 0
>-    "0.0"
>+    "0"
>
>Nothing that causes confusion is great.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/cuducos/elm-format-number/pull/2

-- 
Простите за краткость, создано в K-9 Mail.

cuducos commented:
> Maybe decimals shouldn't be part of Locale, then...

Either it's a sarcastic way to comment on my opinion, or I can't get your point here. If it's the second option, would you mind expanding on it?

I really care abou this package and its API, I'm trying my best, I'm really glad you and @rogeriochaves are contributing to this topic. And I would like to have our difference treated in a better way in this discussion.

lovasoa commented:
Wow! That makes sense… a lot of sense! IMHO this is the best solution so far 🍾 🎊 

As I've said before: if you wanna add that to this PR, that's ok. Otherwise I merge it and add the suggestion to my TODO list ; )

Just let me know if you want me to wait before merging…

lovasoa committed:
Use the unicode minus sign "−" instead of "-"

Use U+2212 MINUS SIGN

lovasoa commented:
thanks :smiley:

cuducos commented:
I merged this PR to a secondary branch, and I'll do some edits myself before bringing it to `master` (I'd have to do it either way because of the URLs pointing to your repo) — but soon your contributions will be on `master` ; )
